File: code/dsa/dsa-c/binary-tree/morris-preorder-traversal.c
#include <stdio.h>
#include <stdlib.h>

struct node 
{
    int data;
    struct node *left;
    struct node *right;
};


/*
Helper function that allocates a new node with
the given data and NULL left and right pointers.
*/
struct node * new_node(int data)
{
    struct node * temp = (struct node*) malloc(sizeof(struct node));
    temp->data = data;
    temp->left = temp->right = NULL;
    return temp;
};


// Preorder traversal without recursion and without stack
void moris_traversal_prorder(struct node * root )
{
    while (root) {

        // If left child is null, print the current node data.
        // Move to right child
        if (root->left == NULL) 
        {
            printf("%d ", root->data);
            root = root->right;
        } else 
        {
            // Find inorder predecessor
            struct node * current = root->left;
            while (current->right && current->right != root) 
                current = current->right;
            
            // If the right child of inorder predecessor already
            // points to this node
            if (current->right == root)
            {
                current->right = NULL;
                root = root->right;
            }

            // if right child doesn't point to this node,
            // then print this node and make right child 
            // point to this node
            else
            {
                printf("%d ", root->data);
                current->right = root;
                root = root->left;
            }
        }
    }
}

// Function for Standrd preorder traversal
void preorder(struct node * root)
{
    if (root)
    {
        printf("%d ", root->data);
        preorder(root->left);
        preorder(root->right);
    }
}


/* Driver program to test above functions*/
int main() 
{ 
    struct node* root = NULL; 
  
    root = newNode(1); 
    root->left = newNode(2); 
    root->right = newNode(3); 
  
    root->left->left = newNode(4); 
    root->left->right = newNode(5); 
  
    root->right->left = newNode(6); 
    root->right->right = newNode(7); 
  
    root->left->left->left = newNode(8); 
    root->left->left->right = newNode(9); 
  
    root->left->right->left = newNode(10); 
    root->left->right->right = newNode(11); 
  
    morrisTraversalPreorder(root); 
  
    printf("\n"); 
    preorder(root); 
  
    return 0; 
} 

File: code/dsa/dsa-py/ds/tree/morris_preorder_traversal.py
from typing import Self


# A binary tree Node
class Node:
    def __init__(self, data) -> None:
        self.data = data
        self.left: Self = None
        self.right: Self = None
        

# Preorder traversal without
# recursion and without stack
def morris_traversal(root: Node):
    curr: Node = root
    
    while curr:
        # If left child is null, print the 
        # current node data, And, update
        # the current point to right child
        if curr.left is None:
            print(curr.data, end=" ")
            curr = curr.right
            
        else:
            # Find the inroder predecessor
            pre: Node = curr.left
            
            while pre.right is not None and pre.right is not curr:
                pre = pre.right
            
            # If the right child of inorder 
            # predecessor already points to 
            # the current node, update the
            # current with it's right child
            if pre.right is curr:
                pre.right = None
                curr = curr.right
                
            # else If right child doesn't point
            # to the current node, then print this
            # node's data and update the right child
            # pointer with the current node and update
            # the current with it's left child
            else:
                print(curr.data, end=" ")
                pre.right = curr
                curr = curr.left

# Function for Standard preorder traversal
def preorder(root: Node):
    if root:
        print(root.data, end=" ")
        preorder(root.left)
        preorder(root.right)
        

if __name__ == "__main__":
    root = Node(1)
    root.left = Node(2) 
    root.right = Node(3) 
    
    root.left.left = Node(4) 
    root.left.right = Node(5) 
    
    root.right.left= Node(6) 
    root.right.right = Node(7) 
    
    root.left.left.left = Node(8) 
    root.left.left.right = Node(9) 
    
    root.left.right.left = Node(10) 
    root.left.right.right = Node(11) 
    
    morris_traversal(root)
    print("\n")
    preorder(root)